File: src/microGauss.jl
#=
miniGauss.jl:
- Julia version: 1.5.1
- Author: Patrizia Favaron
- Date: 2020-10-27
=#

# Main program for 'microGauss' toy Gaussian dispersion model.

using ConfParser
using DataFrames
using CSV

# Get command line parameters
num_args = length(ARGS)
if num_args != 1
    println("microGauss.jl - Toy Gaussian dispersion model")
    println("")
    println("Usage:")
    println("")
    println("  julia microGauss.jl <config_file>")
    println("")
    println("This is open-source software, covered by the MIT License")
    println("")
    println("Written by: Patrizia Favaron")
    println("")
    exit()
end
ini_file = ARGS[1]

# Get configuration data
try
    config = ConfParse(ini_file)
    parse_conf!(config)
    sRunName = retrieve(config, "General", "Run")
    sOutFile = retrieve(config, "General", "OutFile")
    sRunName = retrieve(config, "General", "Run")
    rX0      = retrieve(config, "Receptors", "X0")
    rY0      = retrieve(config, "Receptors", "Y0")
    iNx      = retrieve(config, "Receptors", "NX")
    iNy      = retrieve(config, "Receptors", "NY")
    rDxy     = retrieve(config, "Receptors", "DXY")
    rXe      = retrieve(config, "Emission", "Xe")
    rYe      = retrieve(config, "Emission", "Ye")
    rHe      = retrieve(config, "Emission", "He")
    rVe      = retrieve(config, "Emission", "Ve")
    rDe      = retrieve(config, "Emission", "De")
    rTe      = retrieve(config, "Emission", "Te")
    sMeteo   = retrieve(config, "Meteorology", "MeteoFile")
catch
    println("miniGauss:: Error: Configuration file is missing or invalid")
    exit(1)
end

# Get meteo data as a data-frame
try
    dfMeteo = CSV.read(sMeteo)
catch
    println("miniGauss:: Error: Meteo file is missing or invalid")
    exit(2)
end

File: src/microGauss.f90
! microGauss - An experiment-oriented simplified Gaussian dispersion model
!
! This is open-source code, covered by the MIT license.
!
! Author: Patrizia Favaron
!
program microGauss

    use config
    
    implicit none
    
    ! Locals
    integer             :: iRetCode
    character(len=256)  :: sIniFile
    type(ConfigType)    :: tCfg
    
    ! Get command line parameters
    if(command_argument_count() /= 1) then
        print *, "microGauss - An experiment-oriented simplified"
        print *, "             Gaussian dispersion model"
        print *
        print *, "Usage:"
        print *
        print *, "  ./microGauss <Ini_File>"
        print *
        print *, "This is open-source software, covered by the MIT license."
        print *
        print *, "Author: Patrizia Favaron"
        print *
        stop
    end if
    call get_command_argument(1, sIniFile)
    
    ! Get configuration data
    iRetCode = tCfg % get(sIniFile)
    if(iRetCode /= 0) then
        print *, "microGauss:: error: Invalid configuration - Ret.code = ", iRetCode
        stop
    end if
    
    ! Complete and check configuration
    iRetCode = tCfg % complete()
    if(iRetCode == 1) then
        print *, "microGauss:: error: Source is outside the receptors grid"
        stop
    elseif(iRetCode == 2) then
        print *, "microGauss:: error: Receptors gridis empty"
        stop
    end if
    
    ! Leave
    print *, "*** End Job ***"

end program microGauss
